File: core/captcha/captcha.py
import uuid
import random
import io
import base64
from PIL import Image, ImageDraw, ImageFont
from typing import Optional
from pydantic import BaseModel
from config.settings import settings


class CaptchaResult(BaseModel):
    captcha_base64: str
    captcha_id: str
    captcha_code: Optional[str] = None


class CaptchaService:
    EXPIRE_SECONDS = 300

    def __init__(self, prefix: str):
        self._prefix = prefix
        self._redis = None

    def init(self, redis):
        self._redis = redis

    async def get_captcha(self) -> CaptchaResult:
        captcha_id = str(uuid.uuid4())
        code = ''.join(random.choices('ABCDEFGHJKLMNPQRSTUVWXYZ23456789', k=4))

        img = Image.new('RGB', (100, 38), color=(255, 255, 255))
        draw = ImageDraw.Draw(img)

        try:
            font = ImageFont.truetype("arial.ttf", 24)
        except IOError:
            font = ImageFont.load_default()

        for i, char in enumerate(code):
            x = 10 + i * 22
            y = random.randint(5, 15)
            color = (random.randint(0, 100), random.randint(0, 100), random.randint(0, 100))
            draw.text((x, y), char, font=font, fill=color)

        for _ in range(10):
            x1, y1 = random.randint(0, 100), random.randint(0, 38)
            x2, y2 = random.randint(0, 100), random.randint(0, 38)
            draw.line([(x1, y1), (x2, y2)], fill=(random.randint(0, 200), random.randint(0, 200), random.randint(0, 200)), width=1)

        buffer = io.BytesIO()
        img.save(buffer, format='PNG')
        buffer.seek(0)

        image_base64 = base64.b64encode(buffer.read()).decode('utf-8')

        if self._redis:
            await self._redis.setex(f"{self._prefix}{captcha_id}", self.EXPIRE_SECONDS, code)

        result = CaptchaResult(captcha_base64=f"data:image/png;base64,{image_base64}", captcha_id=captcha_id)
        if settings.app.debug:
            result.captcha_code = code

        return result

    async def check_captcha(self, captcha_id: str, captcha_code: str) -> bool:
        if not captcha_id or not captcha_code:
            raise ValueError("验证码ID或验证码内容不能为空")

        if not self._redis:
            return True

        key = f"{self._prefix}{captcha_id}"
        stored_code = await self._redis.get(key)

        if not stored_code:
            raise ValueError("验证码已过期或无效")

        if stored_code.upper() != captcha_code.strip().upper():
            raise ValueError("验证码错误")

        await self._redis.delete(key)
        return True


b_captcha = CaptchaService(prefix="b:captcha:")
c_captcha = CaptchaService(prefix="c:captcha:")

File: core/captcha/captcha.go
package captcha

import (
	"context"
	"image"
	"image/color"
	"image/draw"
	"image/png"
	"math/rand"
	"strings"

	"bytes"
	"encoding/base64"
	"fmt"
	"time"

	"github.com/redis/go-redis/v9"
	"golang.org/x/image/font"
	"golang.org/x/image/font/basicfont"
	"golang.org/x/image/math/fixed"

	"hei-gin/core/enums"
)

const expireSeconds = 300

// CaptchaResult represents the CAPTCHA response sent to the client.
type CaptchaResult struct {
	CaptchaBase64 string `json:"captcha_base64"`
	CaptchaID     string `json:"captcha_id"`
	CaptchaCode   string `json:"captcha_code,omitempty"`
}

// Service handles CAPTCHA generation and verification.
type Service struct {
	prefix string
	rdb    *redis.Client
}

// NewCaptchaService creates a CAPTCHA service for the given login type prefix.
func NewCaptchaService(loginType string) *Service {
	prefix := strings.ToUpper(loginType) + ":captcha:"
	return &Service{prefix: prefix}
}

// NewBusinessCaptcha creates a CAPTCHA service for BUSINESS login type.
func NewBusinessCaptcha() *Service {
	return NewCaptchaService(string(enums.LoginTypeBusiness))
}

// NewConsumerCaptcha creates a CAPTCHA service for CONSUMER login type.
func NewConsumerCaptcha() *Service {
	return NewCaptchaService(string(enums.LoginTypeConsumer))
}

// Init sets the Redis client.
func (s *Service) Init(rdb *redis.Client) {
	s.rdb = rdb
}

// Generate creates a CAPTCHA image and stores the code in Redis.
func (s *Service) Generate(debug bool) (*CaptchaResult, error) {
	captchaID := fmt.Sprintf("%d", time.Now().UnixNano())
	code := generateCode()

	img := createCaptchaImage(code)

	var buf bytes.Buffer
	if err := png.Encode(&buf, img); err != nil {
		return nil, err
	}
	b64 := base64.StdEncoding.EncodeToString(buf.Bytes())

	if s.rdb != nil {
		ctx := context.Background()
		if err := s.rdb.Set(ctx, s.prefix+captchaID, code, expireSeconds*time.Second).Err(); err != nil {
			return nil, err
		}
	}

	result := &CaptchaResult{
		CaptchaBase64: "data:image/png;base64," + b64,
		CaptchaID:     captchaID,
	}
	if debug {
		result.CaptchaCode = code
	}
	return result, nil
}

// Verify checks a CAPTCHA code against the stored value, then deletes it (one-time use).
func (s *Service) Verify(captchaID, captchaCode string) error {
	if captchaID == "" || captchaCode == "" {
		return fmt.Errorf("验证码ID或验证码内容不能为空")
	}
	if s.rdb == nil {
		return nil
	}

	ctx := context.Background()
	key := s.prefix + captchaID
	stored, err := s.rdb.Get(ctx, key).Result()
	if err == redis.Nil {
		return fmt.Errorf("验证码已过期或无效")
	}
	if err != nil {
		return err
	}

	if !strings.EqualFold(stored, captchaCode) {
		return fmt.Errorf("验证码错误")
	}

	s.rdb.Del(ctx, key)
	return nil
}

func generateCode() string {
	const charset = "ABCDEFGHJKLMNPQRSTUVWXYZ23456789"
	rng := rand.New(rand.NewSource(time.Now().UnixNano()))
	code := make([]byte, 4)
	for i := range code {
		code[i] = charset[rng.Intn(len(charset))]
	}
	return string(code)
}

func createCaptchaImage(code string) image.Image {
	img := image.NewRGBA(image.Rect(0, 0, 100, 38))
	draw.Draw(img, img.Bounds(), &image.Uniform{color.White}, image.Point{}, draw.Src)

	rng := rand.New(rand.NewSource(time.Now().UnixNano()))

	face := basicfont.Face7x13

	for i, ch := range code {
		x := 10 + i*22
		y := 15 + rng.Intn(10)
		c := color.RGBA{
			R: uint8(rng.Intn(100)),
			G: uint8(rng.Intn(100)),
			B: uint8(rng.Intn(100)),
			A: 255,
		}
		drawChar(img, x, y, ch, c, face)
	}

	for range 10 {
		x1 := rng.Intn(100)
		y1 := rng.Intn(38)
		x2 := rng.Intn(100)
		y2 := rng.Intn(38)
		drawLine(img, x1, y1, x2, y2, color.RGBA{
			R: uint8(rng.Intn(200)),
			G: uint8(rng.Intn(200)),
			B: uint8(rng.Intn(200)),
			A: 255,
		})
	}

	return img
}

func drawChar(img *image.RGBA, x, y int, ch rune, c color.Color, face font.Face) {
	d := &font.Drawer{
		Dst:  img,
		Src:  image.NewUniform(c),
		Face: face,
		Dot: fixed.Point26_6{
			X: fixed.Int26_6(x * 64),
			Y: fixed.Int26_6(y * 64),
		},
	}
	d.DrawString(string(ch))
}

func drawLine(img *image.RGBA, x1, y1, x2, y2 int, c color.Color) {
	dx := x2 - x1
	dy := y2 - y1
	steps := max(abs(dx), abs(dy))
	if steps == 0 {
		return
	}
	for i := 0; i <= steps; i++ {
		x := x1 + dx*i/steps
		y := y1 + dy*i/steps
		if x >= 0 && x < 100 && y >= 0 && y < 38 {
			img.Set(x, y, c)
		}
	}
}

func abs(x int) int {
	if x < 0 {
		return -x
	}
	return x
}
